service/matching/incoming_service_manager.go

```diff
@@ -0,0 +1,231 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package matching
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"github.com/google/uuid"
+
+	"go.temporal.io/server/api/matchingservice/v1"
+	persistencepb "go.temporal.io/server/api/persistence/v1"
+	"go.temporal.io/server/common/clock"
+	hlc "go.temporal.io/server/common/clock/hybrid_logical_clock"
+	"go.temporal.io/server/common/nexus"
+	p "go.temporal.io/server/common/persistence"
+)
+
+const (
+	// loadServicesPageSize is the page size to use when initially loading services from persistence
+	loadServicesPageSize = 100 // TODO: should be dynamic config?
+)
+
+type (
+	// incomingServiceManager manages cache and persistence access for Nexus incoming services.
+	// incomingServiceManager also contains a RWLock to enforce serial updates to prevent
+	// nexus_incoming_services table version conflicts.
+	//
+	// incomingServiceManager should only be used within matching service because it assumes
+	// that it is running on the matching node that owns the nexus_incoming_services table.
+	// There is no explicit listener for membership changes because table ownership changes
+	// will be detected by version conflicts and eventually settle through retries.
+	//
+	// Not to be confused with persistence.NexusServiceManager which is responsible for persistence-layer
+	// CRUD APIs for Nexus incoming services.
+	incomingServiceManager interface {
+		CreateOrUpdateNexusIncomingService(ctx context.Context, request *matchingservice.CreateOrUpdateNexusIncomingServiceRequest, clusterID int64, timeSource clock.TimeSource) (*matchingservice.CreateOrUpdateNexusIncomingServiceResponse, error)
+		DeleteNexusIncomingService(ctx context.Context, request *matchingservice.DeleteNexusIncomingServiceRequest) (*matchingservice.DeleteNexusIncomingServiceResponse, error)
+		ListNexusIncomingServices(ctx context.Context, request *matchingservice.ListNexusIncomingServicesRequest) (*matchingservice.ListNexusIncomingServicesResponse, chan struct{}, error)
+	}
+
+	incomingServiceManagerImpl struct {
+		hasLoadedServices atomic.Bool
+
+		sync.RWMutex        // protects tableVersion, servicesByName, and tableVersionChanged
+		tableVersion        int64
+		servicesByName      map[string]*persistencepb.NexusIncomingServiceEntry
```

Comment:
How are you dealing with the non-deterministic iteration order here?
I think you'll probably want a different data structure.
It could be as simple as a slice and a map that indexes by name.

Reply:
Yeah, there is a TODO further down where I mention this. The main issue is that we use the `service_id` as the page token, but all of our access is by name. The straightforward solution is to maintain a sorted list of IDs, a map of ID -> name and a map of name -> service. That's a big ugly, so I'm still trying to think of something more elegant.

Reply:
You could use a treemap like we do for matching's ackManager; that'll give you a KV data structure with deterministic iteration order

Reply:
That'd also let you skip the binary search of your services slice; you'd just do `myTreemap.Put(id, entry)`

Reply:
As long as you can maintain the iteration order that you'd get from the DB (ID ASC) and also have a "secondary index" where you can look up by name, I don't mind what you go with.

I suggested a list with entries coming from the DB and a map of service name to service.

Reply:
I don't think a treemap will quite work. It orders elements by key, so we would have to use service ID as a key to match the order in the database. Users will only reference services by name, not ID, so we would still have to maintain a second data structure to map name -> ID anyway. I know the binary search isn't ideal, but I think the performance hit shouldn't be too significant given we don't expect these APIs to be very hot paths, especially as compared to resolving service name -> namespace + task queue.